### ISSUE TITLE
Can't login after switch to messenger.com urls

### ISSUE THREAD

**Issue by wetmore** [carpedm20/fbchat#582]
## Description of the problem
Attempting to login using fbchat v2.0.0a3 throws an error, see Traceback below. I can login as expected using 2.0.0a2 and 1.9.6..

Someone else has had the same problem, and could also solve by using v2.0.0a2, see https://github.com/carpedm20/fbchat/issues/581

Seems related to the solution to https://github.com/carpedm20/fbchat/issues/413, in particular commit https://github.com/carpedm20/fbchat/commit/079d4093c4f0febf004c6cf8f6972f157833a4f7

## Code to reproduce
```py
import fbchat

session = fbchat.Session.login("<username snipped>", "<password snipped>")
```

## Traceback
```
Traceback (most recent call last):
  File "hello.py", line 4, in <module>
    session = fbchat.Session.login("<username snipped>", "<password snipped>")
  File "/Users/matt/programming/python/patrick2/venv/lib/python3.7/site-packages/fbchat/_session.py", line 290, in login
    raise _exception.NotLoggedIn(error)
fbchat.NotLoggedIn: Password Forgot account?
```

## Environment information
- Python version 3.7.7
- `fbchat` version 2.0.0a3

**Comment by wetmore** [comment]
I did a little poking around. I suppose the more relevant commit here is https://github.com/carpedm20/fbchat/commit/7be2acad7d8cf3e05cb3676c828e0695df38bbd9, which adds the code which throws the error:

```
# We weren't redirected, hence the email or password was wrong
if not url:
    error = get_error_data(r.content.decode("utf-8"))
    raise _exception.NotLoggedIn(error)
```

In my case, instead of being redirected to the messenger UI (302), the login request gets a response with code 200, and the content of the response is a big blob of html which renders as:

![response content](https://user-images.githubusercontent.com/719776/82951068-81261200-9f74-11ea-902b-e498ddddb554.png)

So you can see the error message which is getting parsed out of it ("Password Forgot account?") is just text from the login bar at the top. The error message returned from FB in the picture indicates the issue is actually due to the request being sent.

When I actually go to messenger.com in a browser and log in with my email and password, it's successful. This indicates to me that the problem will be present in the diff between the request that fbchat sends, and the request that messenger.com sends when I login using that page. One option going forwards for me would be to start experimenting with changing the headers and data sent in the login request. However, it seems strange to me that I would have this problem but someone else would be able to login through fbchat just fine. 

Can anyone confirm that they can log in using 2.0.0a3? I'm assuming you can @madsmtm.

**Comment by xaadu** [comment]
I can't login! I've tried today! Exact same problem! @wetmore

**Comment by wetmore** [comment]
> I can't login! I've tried today! Exact same problem! @wetmore

In the meantime I am using 2.0.0a2 and it's working fine, can you use that workaround?

**Comment by xaadu** [comment]
I've tried login with 2a before and I was success! But 3a doesn't work for
me.

**Comment by madsmtm** [comment]
Thanks for the report!

I can't reproduce it myself, so maybe one of you would like to help me fix it? If so, please follow these instructions:
- Go to [https://www.messenger.com/login](https://www.messenger.com/login) in your browser
- Open your developer tools on the "Inspect" tab
- Find the `form` HTML element in there
- Inside this element, you should see a bunch of hidden `input` HTML elements, with different `name`s and `value`s. Note these down.
- Go to [Line 272 in `_session.py`](https://github.com/carpedm20/fbchat/blob/520258e339c1826fe28649fbc3aee3afa1117a67/fbchat/_session.py#L272-L285), and try inputting different values, matching the ones you found in the previous step.
- After a bit of trial and error, you might be able to find out which of these values are required for you to not hit the error.

It's quite a manual approach, and also probably quite hard to follow, so if something here doesn't work, please ask me to elaborate

**Comment by xaadu** [comment]
I've been trying for 2 hours and nothing happened! I've tried with sending every single details, but it can't log in.

I've added dynamic details to the session so that every details gets up to date perfectly
`d=session.get('https://messenger.com/login').text
inputs=bs4.BeautifulSoup(d, 'html.parser').find_all('input')
data = {
            "jazoest": inputs[0]["value"],
            "lsd": inputs[1]["value"],
            "initial_request_id": inputs[2]["value"],  # any, just has to be present
            "timezone": "-360",
            "lgndim": inputs[4]["value"],
            "lgnrnd": inputs[5]["value"],
            "lgnjs": inputs[6]["value"],
            "email": email,
            "pass": password,
            "login": "1",
            "persistent": inputs[9]["value"],  # Changes the cookie type to have a long "expires"
            "default_persistent": inputs[10]["value"],
        }`

Still, can't login! NB: value of the key "lgndim" gets empty string, don't know why, so I've tried adding custom value to that, still doesn't work! (Maybe it is generated after the page load using JS)

After trying to log in many times, I've found that, a cookie generates and then logs in to messenger named "act" followed by a number value, and that is the main problem. I have no idea where the cookie sent from! 
Here's an example: 
After going to login page: the cookies are - 
`'cookie': '_fbp=fb.1.1591773315379.820061979; wd=788x625; datr=g4jgXjB4k6l4gPODUwdd4Rt9'`
and this cookie is sent by browser on login attempt -
`'cookie': '_fbp=fb.1.1591773315379.820061979; wd=788x625; datr=g4jgXjB4k6l4gPODUwdd4Rt9; act=1591773579064%2F1'`

See, last act just appeared! When this cookie is available and I send request to login, it goes to the 2fa page, but if the act is not available in the cookie, then It redirects to the error page @wetmore  provided! I'm noob at coding so I don't know what happens here! I've told you everything I could feel about this! And one more thing, the act numbers first some digits matches with the "datr" cookie and more matches with the max-age or something from received headers strict-transport-security maybe! I'm not sure! 

I've explained everything I could get on this login, I'll try to explore more If I get free time! I don't know if anything I wrote could help! Please tell me if i should try another way or something!

**Comment by madsmtm** [comment]
Thanks for the details @xaadu, they've been really helpful!

Like you said, the `act` cookie is generated using JavaScript (I've seen it as well 😉). To me, it seems to be a combination of the current timestamp (in milliseconds) and a number indicating which number request this was (we set this to `0` always, might not be correct?).

I've added this to the login code, see d65094653154ebecc5f3fd018164a0459b5d259c, and released `v2.0.0a5`, so please try that out and see if it works for you now.

**Comment by madsmtm** [comment]
Note to my future self: I managed to get the same error page as you people by inserting an old `datr` cookie, using that I should be able to improve the error message parsing.

**Comment by xaadu** [comment]
Same problem bro! ☹☹ 

![image](https://user-images.githubusercontent.com/41540861/84614012-a1e3e680-aee6-11ea-9598-40e4c4b8c4c9.png)

**Comment by xaadu** [comment]
I've found a way to log in using pyppeteer! I've logged in and got the session perfectly! It sometimes hangs because of some function call! But it works most of the time! Should I polish the code? Or it can't be used because of weight? I didn't give any fallback for the wrong username/pass/2fa code! I'll polish if I'm sure that pyppeteer is usable! Here's the code (Almost unreadable noob code for the testing purpose that needs to be polished perfectly):

```python
import asyncio
from pyppeteer import launch
import JSON


email='antheaxs3'
password='xadu@xaadu'


async def main():
    browser = await launch()
    page = await browser.newPage()
    await page.goto('https://messenger.com/login', { 'waitUntil': 'load' })
    #await page.waitForSelector('#u_0_0')
    try:
        await page.click('#u_0_0')
    except:
        await asyncio.sleep(1)
        await page.click('#u_0_0')
    await page.click('#email')
    await page.keyboard.sendCharacter(email)
    await page.click('#pass')
    await page.keyboard.sendCharacter(password)
    await page.keyboard.press('Enter')
    await page.waitForNavigation({ 'waitUntil': 'load' })
    print(page.url)
    while 'checkpoint_interstitial' in page.url:
        await page.click('#XMessengerDotComLoginViewPlaceholder > div > div._3-mr > div > a', { 'waitUntil': 'load' })
        #XMessengerDotComLoginViewPlaceholder > div > div._3-mr > div > a
        await page.screenshot({'path': 'example.png'})
        try:
            await page.setJavaScriptEnabled(False)
            await page.JJ('#u_0_c')
            await page.click('#approvals_code')
            code=input('Enter Code for 2FA Authentication: ')
            await page.keyboard.sendCharacter(code)
            await page.screenshot({'path': 'example.png'})
            await page.click('#checkpointSubmitButton', { 'waitUntil': 'load' })
            await page.screenshot({'path': 'example.png'})
            await page.waitForSelector('#u_0_3')
            await page.click('#u_0_3')
            await page.waitForSelector('#checkpointSubmitButton')
            await page.click('#checkpointSubmitButton', { 'waitUntil': 'load' })
            await page.screenshot({'path': 'example.png'})
            raise Exception
        except Exception as e:
            await page.waitForSelector('#checkpointSubmitButton')
            await page.click('#checkpointSubmitButton')
            await page.screenshot({'path': 'example.png'})
            await page.waitForSelector('#checkpointSubmitButton')
            await page.click('#checkpointSubmitButton')
            await page.screenshot({'path': 'example.png'})
            await page.waitForSelector('#checkpointSubmitButton')
            await page.setJavaScriptEnabled(True)
            await page.click('#checkpointSubmitButton', { 'waitUntil': 'load' })
            await page.screenshot({'path': 'example.png'})
            print('Done')

    cookies=await page.cookies()
    with open('test.json', 'w') as f:
        f.write(json.dumps(cookies))
    finalCookies={}
    for cookie in cookies:
        finalCookies[cookie['name']]=cookie['value']
    with open('session.json', 'w') as f:
        f.write(json.dumps(finalCookies))

    print(finalCookies)

    await page.screenshot({'path': 'example.png'})
    await browser.close()

asyncio.get_event_loop().run_until_complete(main())
```

Note: This will automatically download the required browser on the first run. So I think it would be supported in any hosting! Selenium is easy but need buildpack in Heroku! 

@madsmtm kindly just answer if it's possible to use pyppeteer in fbchat! It's a temporary fix!

**Comment by TimLChan** [comment]
The code above is a bit bloated to be committed I reckon. You're better off just scraping the `_js_datr` cookie from a new session on `messenger.com`.

Add it to your session cookies using:
`requests.utils.add_dict_to_cookiejar(session.cookies, {'_js_datr': js_datr})`
or
`session.cookies.set("_js_datr", js_datr, domain=".messenger.com")`

**Comment by xaadu** [comment]
Thanks, @TimLChan . I've fixed it the same way you've mentioned but I didn't push in this repo because I was sure It was not gonna work and now it doesn't. All credit goes to Facebook for changing the login system 😑😑😑

So I think the best is to use "pyppeteer" based login system. ☹

**Comment by TimLChan** [comment]
There's some other session variables that need to get fixed for it to work which I've done on my branch. I haven't written any tests or thoroughly tested them so I haven't raised a PR.

**Comment by xaadu** [comment]
Wow! Nice work! Do your code still work? Cause mine doesn't.

**Comment by shashankredemption** [comment]
@TimLChan does your fork still work with the new messenger/ig merge?

**Comment by jgordonpds** [comment]
I'm actively trying to fix this issue in my fork. Anyone have any updates regarding this?
